persistence-jpa/javadsl/src/main/java/com/lightbend/lagom/internal/javadsl/persistence/jpa/JpaSessionImpl.java

```diff
@@ -0,0 +1,134 @@
+        return factoryCompletionStage.thenCompose(factory -> executeInSlickContext(() -> {
+            EntityManager entityManager = factory.createEntityManager();
+            try {
+                return block.apply(entityManager);
+            } finally {
+                entityManager.close();
+            }
+        }));
+    }
+
+    @Override
+    public <T> CompletionStage<T> withTransaction(Function<EntityManager, T> block) {
+        return withEntityManager(entityManager -> {
+            EntityTransaction transaction = entityManager.getTransaction();
+            try {
+                transaction.begin();
+                T result = block.apply(entityManager);
+                transaction.commit();
+                return result;
+            } finally {
+                if (transaction.isActive()) transaction.rollback();
+            }
+        });
+    }
+
+    private CompletionStage<EntityManagerFactory> tryCreateEntityManagerFactory(Deadline deadline,
+                                                                                FiniteDuration retryInterval) {
+        CompletionStage<EntityManagerFactory> result = executeInSlickContext(() ->
+                Persistence.createEntityManagerFactory(PERSISTENCE_UNIT_NAME)
+        );
+        // This wrapping and then composing with identity logic is because
+        // CompletionStage has no equivalent to Scala's Future.recoverWith
+        return result.thenApply(CompletableFuture::completedFuture)
+                .exceptionally(e -> retryCreateEntityManagerFactory(e, deadline, retryInterval))
+                .thenCompose(Function.identity());
+    }
+
+    private CompletableFuture<EntityManagerFactory> retryCreateEntityManagerFactory(Throwable exception,
+                                                                                    Deadline deadline,
+                                                                                    FiniteDuration retryInterval) {
+        log.warn("Exception while initializing JPA EntityManagerFactory", exception);
+        CompletableFuture<EntityManagerFactory> future = new CompletableFuture<>();
+        if (deadline.timeLeft().lteq(retryInterval)) {
+            // Duration is up, don't try and recover from failure
+            future.completeExceptionally(exception);
+        } else {
+            log.info("Will retry initializing JPA EntityManagerFactory in {}", retryInterval);
+            actorSystem.scheduler().scheduleOnce(
+                    retryInterval,
+                    () -> tryCreateEntityManagerFactory(deadline, retryInterval)
+                            .thenAccept(future::complete)
+                            .exceptionally(ex -> {
+                                future.completeExceptionally(ex);
+                                return null;
+                            }),
+                    actorSystem.dispatcher()
+            );
+        }
+        return future;
+    }
```

Comment:
Heh, well I did a little more Google searching and found these:

- https://issues.scala-lang.org/browse/SI-8615
- https://gist.github.com/viktorklang/9414163

That uses a retry count rather than duration, but that might make more sense anyway.